src/main/java/com/iota/iri/service/milestone/MilestoneSolidifier.java

```diff
@@ -0,0 +1,247 @@
+     */
+    private TransactionValidator transactionValidator;
+
+    /**
+     * List of unsolid milestones where we collect the milestones that shall be solidified.
+     */
+    private ConcurrentHashMap<Hash, Integer> unsolidMilestones = new ConcurrentHashMap<>();
+
+    /**
+     * The earliest unsolid milestone hash which is the one that is actively tried to get solidified.
+     */
+    private Hash earliestMilestoneHash = null;
+
+    /**
+     * The earliest unsolid milestone index which is the one that is actively tried to get solidified.
+     */
+    private int earliestMilestoneIndex = Integer.MAX_VALUE;
+
+    /**
+     * Holds the amount of solidity checks issued for the earliest milestone.
+     */
+    private int earliestMilestoneTries = 0;
+
+    /**
+     * A flag indicating if the solidifier thread is running.
+     */
+    private boolean running = false;
+
+    /**
+     * Constructor of the class.
+     *
+     * It simply stores the passed in parameters to be able to access them later on.
+     *
+     * @param snapshotManager SnapshotManager instance that is used by the node
+     * @param transactionValidator TransactionValidator instance that is used by the node
+     */
+    public MilestoneSolidifier(SnapshotManager snapshotManager, TransactionValidator transactionValidator) {
+        this.snapshotManager = snapshotManager;
+        this.transactionValidator = transactionValidator;
+    }
+
+    /**
+     * This method adds new milestones to our internal map of unsolid ones.
+     *
+     * It first checks if the passed in milestone is relevant for our node by checking against the initialSnapshot. If
+     * it deems the milestone relevant it checks if the passed in milestone appeared earlier than the current one and
+     * updates the interval properties accordingly.
+     *
+     * @param milestoneHash Hash of the milestone that shall be solidified
+     * @param milestoneIndex index of the milestone that shall be solidified
+     */
+    public void add(Hash milestoneHash, int milestoneIndex) {
+        if (milestoneIndex > snapshotManager.getInitialSnapshot().getIndex()) {
+            if (milestoneIndex < earliestMilestoneIndex) {
+                earliestMilestoneHash = milestoneHash;
+                earliestMilestoneIndex = milestoneIndex;
+                earliestMilestoneTries = 0;
+            }
+
+            unsolidMilestones.put(milestoneHash, milestoneIndex);
+        }
+    }
+
+    /**
+     * This method starts the solidification thread that periodically updates and checks the unsolid milestones.
+     *
+     * To allow for faster processing it only "waits" for another check if the current solidification task was not
+     * finished (otherwise it immediately "continues" with the next one).
+     */
+    public void start() {
+        running = true;
+
+        new Thread(() -> {
+            while(running) {
+                if (processSolidificationTask() && running) {
```

Comment:
You are right - doesnt make much sense before i had some long running method in the if clause that i wanted to prevent from getting executed if the thread was terminated while the solidification took place